Support for CONNECT BY

jmhain opened the pull request:
This adds support for the [`CONNECT BY`](https://docs.snowflake.com/en/sql-reference/constructs/connect-by) operator, supported by Snowflake, Redshift, and MS SQL.

A bit of awkwardness is the `PRIOR` expressions used to refer to the upper level in the `CONNECT BY` relationships. These expressions are only valid in this context, and `prior` is a perfectly valid identifier elsewhere. Trying to thread that context through the call stack was resulting in a massive diff, so I added some state to the parser to avoid that. Happy to change this is reviewers are not fond of that approach.

alamb commented on src/ast/query.rs:
```diff
@@ -92,6 +92,8 @@ impl fmt::Display for Query {
     /// Restricted SELECT .. FROM .. HAVING (no ORDER BY or set operations)
     Select(Box<Select>),
+    /// SELECT .. FROM .. STARTING WITH .. CONNECT BY
+    ConnectBy(ConnectBy),
```
Your rationale from the PR description is quite well stated and understandable and I think would be a good addition to the code

```suggestion
    /// 
    /// Note: `ConnectBy` is variant of `SetExpr`, rather than `TableFactor` or a field on `Select` 
    /// because, despite being part of a `SELECT .. FROM` clause, `CONNECT BY` does not compose
    /// with other table factors (e.g. `PIVOT`) in a `FROM`, nor does it permit most options usually available
    ///  in a `SELECT` (e.g. adding a `WHERE` clause in a `SELECT .. CONNECT BY` yields a parse error 
    /// from Snowflake).
    ConnectBy(ConnectBy),
```

alamb commented on src/parser/mod.rs:
```diff
@@ -6936,6 +6965,48 @@ impl<'a> Parser<'a> {
+        })))
+    }
+
+    fn with_state<T, F>(&mut self, state: ParserState, mut f: F) -> Result<T, ParserError>
```
```suggestion
    /// Invoke `f` after first setting the parser's `ParserState` to `state`. 
    ///
    /// Upon return, restores the parser's state to what it started at
    fn with_state<T, F>(&mut self, state: ParserState, mut f: F) -> Result<T, ParserError>
```

alamb commented on tests/sqlparser_snowflake.rs:
```diff
@@ -1173,3 +1173,88 @@ fn parse_top() {
+        vec![SelectItem::UnnamedExpr(Expr::Identifier(Ident::new(
+            "prior"
+        )))]
+    );
```
Can you please add some negative cases as well. Perhaps using your rationale in the description:

1. try to compose with other table factors (e.g. PIVOT) in a FROM,
2. Some options usually available in a SELECT (e.g. adding DISTINCT or WHERE clause in a SELECT .. CONNECT BY)

jmhain replied:
Most of these cases ended up being supported on other dialects, so they're now allowed.

alamb reviewed (commented):
Thank you very much for this contribution @jmhain  -- I am truly sorry for the delay in review. 

I think the approach taken by this PR looks good to me. In my opinion, all that is needed prior to merge is:
1. Some additional comments (I left some suggestions)
2. A few negative test cases

Thanks again for such an easy to read and review PR

alamb commented:
Marking as Draft to signify this PR is no longer waiting on review. Please mark it as ready for review when it is ready for another look.

jmhain commented:
@alamb Many thanks for the review feedback, I greatly appreciate it! And sorry for the delay in getting to this. So I learned that MS SQL and Redshift both also support `CONNECT BY`, and I did some more testing and found that my approach before was really only well suited to Snowflake, so I reworked this significantly. I also added some more tests. Thanks again!

jmhain commented:
Just resolved conflicts on this again cc @alamb

alamb reviewed (approved):
Thanks again @jmhain

alamb commented:
Looks like there is a conflict (perhaps with one of your other PRs 😆 ) -- if you can resolve that I'll merge this PR in.

Thanks again

jmhain commented:
@alamb Done, thank you!

alamb commented:
🚀
